File: 12.整数转罗马数字.go
/*
 * @lc app=leetcode.cn id=12 lang=golang
 *
 * [12] 整数转罗马数字
 */

// @lc code=start
func intToRoman(num int) string {
	type tuple struct { value int; char string }
	tupleList := []tuple{
		tuple{1000, "M"},
		tuple{900, "CM"},
		tuple{500, "D"},
		tuple{400, "CD"},
		tuple{100, "C"},
		tuple{90, "XC"},
		tuple{50, "L"},
		tuple{40, "XL"},
		tuple{10, "X"},
		tuple{9, "IX"},
		tuple{5, "V"},
		tuple{4, "IV"},
		tuple{1, "I"},
	}
	
	ret := []string{};
	for _, t := range tupleList {
		for num >= t.value {
			num -= t.value
			ret = append(ret, t.char)
		}
		if num == 0 {
			break
		}
	}
	return strings.Join(ret, "")
}
// @lc code=end

File: 12.整数转罗马数字.rs
/*
 * @lc app=leetcode.cn id=12 lang=rust
 *
 * [12] 整数转罗马数字
 */

// @lc code=start
impl Solution {
    pub fn int_to_roman(num: i32) -> String {
        let mut num = num;
        let tupleList = vec![
            (1000, "M"),
            (900, "CM"),
            (500, "D"),
            (400, "CD"),
            (100, "C"),
            (90, "XC"),
            (50, "L"),
            (40, "XL"),
            (10, "X"),
            (9, "IX"),
            (5, "V"),
            (4, "IV"),
            (1, "I"),
        ];
        let mut ret: Vec<String> = vec![];
        for (value, symble) in tupleList.into_iter() {
            while num >= value {
                num -= value;
                ret.push(symble.to_string());
            }
            if num == 0 {
                break;
            }
        }
        ret.join("")
    }
}
// @lc code=end

